dask/array/slicing.py

```diff
@@ -984,6 +1012,16 @@ def slice_with_bool_dask_array(x, index):
     """
     from .core import Array, atop, elemwise
 
+    x_chunks = _flat_chunks(x)
+    nan_in_x_chunks = np.isnan(x_chunks).any()
+    index_chunks = _flat_chunks(index)
+    nan_in_index_chunks = np.isnan(index_chunks).any()
+    failure_mode = nan_in_x_chunks and not nan_in_index_chunks
+
+    if failure_mode:
```

Comment:
This check is too strict (see the test failures). We only fail if missing dimensions align with those being indexed. Given `x` is a 2D array with unknown dimensions along the first axis:

- `x[y > 1, :]` should fail
- `x[:, y > 1]` should pass, since we're leaving the unknown axis unchanged

To check if an axis of a dask array is unknown, you can just look at the `.shape` property and see if it's `nan` along that axis.

```python
In [16]: x = da.random.normal(size=(100, 100), chunks=(50, 50))

In [17]: index = x.sum(axis=1) > 100

In [18]: missing = x[index, :]

In [19]: missing.shape
Out[19]: (nan, 100)
```

Reply:
@jcrist thanks for the clarification on what needs to be checked. I was trying to use the tests as the basis for the check's logic, but unfortunately the still-failing tests on Travis were skipping locally so I missed the logic they implied.

I've re-implemented the logic here, which has left only one test that now fails ([this one](https://github.com/dask/dask/blob/master/dask/array/tests/test_array_core.py#L3083)). Based on the logic above, I can't see how this test could now pass in its current form. I could...
* update the test (for example, if the array's chunks are set to `nan` after the two indexing assertions then this test will pass)
* further modify the logic (I might still be missing some required behaviour)

But I'm not sure which is best.

Reply:
For reference, here's [that test failing in Travis](https://travis-ci.org/dask/dask/jobs/451300049#L1466)